File: zk/state_audit.py
"""Zero-knowledge audit of a market maker's *state update*, not just its policy.

`policy_audit.py` shows the pricing rule is well-formed: every field of the rule
sits inside a band the venue published, and the shares the computing nodes hold
open to the committed rule. That covers the rule. It says nothing about what
happens after a fill, and in a stateful protocol --- RFS carries the winner's
inventory forward from one quote to the next --- the rule is the smaller half.
A maker whose declared rule is impeccable can still carry an inventory that
never moved when it should have, or that quietly grew past the size it promised
to stop at.

Three things are proved here, per fill, without opening anything.

    arithmetic   the new inventory is the old one less what was filled, as a
                 linear relation over three commitments
    containment  the new inventory is inside a limit that is itself committed,
                 so the venue learns the promise was kept without learning
                 either the promise or the position
    continuity   each step names the state it followed, so replaying an old
                 state or running two states in parallel breaks the chain

Containment is the part that needs a committed bound rather than a public one.
A public band would have to be wide enough for the largest maker on the venue,
which makes it vacuous for everyone else; the point of a private limit is that
a maker is held to the promise it actually made. The bound is committed once,
with the policy, and every later step proves against that same commitment ---
so tightening the promise after a breach is not available either.

What is deliberately not claimed: this shows the state the maker *published*
evolved correctly. Binding that state to the inventory the MPC nodes actually
used has the same open edge as `policy_audit.py` --- MP-SPDZ works over its own
field --- and closing it is the same piece of work, not a second one.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Sequence

from .commit import (
    OpeningProof, Pedersen, RangeProof, prove_bounded, prove_linear,
    verify_bounded, verify_linear,
)
from .groups import Group


@dataclass(frozen=True)
class InventoryLimit:
    """The bound a maker commits to once, and is held to at every later step."""

    commitment: Any
    range_proof: RangeProof
    ceiling: int                      # the public ceiling the limit itself sits under

    def encoded(self, group: Group) -> bytes:
        return group.encode(self.commitment)


@dataclass(frozen=True)
class StateStep:
    """One link of the inventory chain: the state left behind by one fill."""

    step: int
    inventory: Any                    # commitment to the inventory after the fill
    fill: Any                         # commitment to the signed size that was filled
    follows: bytes                    # encoding of the inventory this step started from
    arithmetic: OpeningProof          # new + filled - old == 0
    below_cap: RangeProof             # limit - new >= 0
    above_floor: RangeProof           # limit + new >= 0


class StateAuditor:
    """Builds and checks the chain. The same object serves both sides.

    Proving needs the openings and verifying does not, which is the only
    difference; keeping them together stops the two sides from drifting apart in
    what they think the context string is, and a mismatched context is the
    failure that looks like a broken proof rather than a broken binding.
    """

    def __init__(self, group: Group, label: bytes = b"qomm:state:v1",
                 ceiling: int = 1 << 20):
        self.group = group
        self.key = Pedersen(group, label)
        self.ceiling = ceiling

    # --- the limit, committed once ---------------------------------------
    def commit_limit(self, limit: int, blinding: int) -> InventoryLimit:
        if not 0 <= limit <= self.ceiling:
            raise ValueError(f"limit {limit} outside [0, {self.ceiling}]")
        commitment, proof, _ = prove_bounded(
            self.key, limit, blinding, 0, self.ceiling, self._context(b"limit"))
        return InventoryLimit(commitment, proof, self.ceiling)

    def check_limit(self, limit: InventoryLimit) -> bool:
        return verify_bounded(self.key, limit.commitment, limit.range_proof,
                              0, limit.ceiling, self._context(b"limit"))

    # --- one step ---------------------------------------------------------
    def prove_update(self, *, step: int, old_inventory: int, old_blinding: int,
                     filled: int, fill_blinding: int, limit: int, limit_blinding: int,
                     new_blinding: int) -> tuple[StateStep, int]:
        """The maker's side. Returns the step and the new inventory value.

        `filled` is signed the way the maker's book moves: a maker that sold
        carries a negative position afterwards, so the new inventory is the old
        one *less* what left. Getting that sign backwards is the mistake this
        proof exists to make impossible to hide, so it is stated once here and
        the relation below is written to match.
        """
        new_inventory = old_inventory - filled
        if abs(new_inventory) > limit:
            raise ValueError(
                f"inventory {new_inventory} breaks the committed limit {limit}; "
                "the maker cannot prove this step and must decline the fill")

        old_commitment = self.key.commit(old_inventory, old_blinding)
        fill_commitment = self.key.commit(filled, fill_blinding)
        new_commitment = self.key.commit(new_inventory, new_blinding)
        context = self._context(f"step:{step}".encode())

        # new + filled - old == 0
        arithmetic = prove_linear(
            self.key, [new_blinding, fill_blinding, old_blinding], [1, 1, -1],
            context, commitments=[new_commitment, fill_commitment, old_commitment])

        below, above = self._containment(
            new_inventory, new_blinding, limit, limit_blinding, context)
        return StateStep(step=step, inventory=new_commitment, fill=fill_commitment,
                         follows=self.group.encode(old_commitment),
                         arithmetic=arithmetic, below_cap=below,
                         above_floor=above), new_inventory

    def _containment(self, inventory: int, blinding: int, limit: int,
                     limit_blinding: int, context: bytes) -> tuple[RangeProof, RangeProof]:
        """|inventory| <= limit, as two one-sided proofs on committed differences.

        The commitment each proof covers is the quotient of two commitments the
        verifier already holds, so passing the difference's own opening here
        produces exactly that point --- checked rather than assumed, because a
        mismatch would make the proof cover a number nobody constrained.
        """
        order = self.group.order
        proofs = []
        for sign in (+1, -1):
            value = limit - sign * inventory
            blind = (limit_blinding - sign * blinding) % order
            commitment, proof, _ = prove_bounded(
                self.key, value, blind, 0, 2 * self.ceiling,
                context + (b":below" if sign > 0 else b":above"))
            expected = self.group.mul(
                self.key.commit(limit, limit_blinding),
                self.group.neg(self.key.commit(sign * inventory % order,
                                               sign * blinding % order)))
            assert self.group.encode(commitment) == self.group.encode(expected)
            proofs.append(proof)
        return proofs[0], proofs[1]

    def verify_update(self, step: StateStep, old_commitment, limit: InventoryLimit) -> bool:
        """The venue's side. Nothing here needs an opening."""
        group = self.group
        if group.encode(old_commitment) != step.follows:
            return False                      # this step did not follow that state
        for point in (step.inventory, step.fill, old_commitment, limit.commitment):
            if not group.is_valid(point):
                return False
        context = self._context(f"step:{step.step}".encode())
        if not verify_linear(self.key,
                             [step.inventory, step.fill, old_commitment], [1, 1, -1],
                             0, step.arithmetic, context):
            return False
        for proof, sign, suffix in ((step.below_cap, +1, b":below"),
                                    (step.above_floor, -1, b":above")):
            difference = group.mul(
                limit.commitment,
                group.neg(group.point_pow(step.inventory, sign % group.order)))
            if not verify_bounded(self.key, difference, proof, 0, 2 * self.ceiling,
                                  context + suffix):
                return False
        return True

    def verify_chain(self, opening_commitment, steps: Sequence[StateStep],
                     limit: InventoryLimit) -> tuple[bool, str]:
        """Walk the chain from a known opening state.

        Returns the reason on failure rather than a bare false, because the three
        ways a chain breaks --- bad arithmetic, a breached limit, a state that
        does not follow the one before it --- call for different responses from
        the venue, and only the last of them is evidence of equivocation.
        """
        if not self.check_limit(limit):
            return False, "the committed limit is not itself in range"
        current = opening_commitment
        for index, step in enumerate(steps):
            if self.group.encode(current) != step.follows:
                return False, (f"step {step.step} (index {index}) does not follow the "
                               "state before it: a replayed or forked inventory")
            if not self.verify_update(step, current, limit):
                return False, f"step {step.step} (index {index}) failed its proofs"
            current = step.inventory
        return True, "ok"

    def _context(self, tag: bytes) -> bytes:
        return b"qomm:state:v1:" + tag

File: rust/qomm-proofs/tests/state_audit.rs
//! The chain has to accept an honest book and reject each way one can be wrong,
//! and the three failures have to stay distinguishable --- only a fork is
//! evidence of equivocation, so a venue that cannot tell them apart cannot act
//! on any of them.

use curve25519_dalek::scalar::Scalar;
use qomm_proofs::state_audit::{ChainError, StateAuditor, StateStep};
use rand_core::OsRng;

struct Book {
    auditor: StateAuditor,
    limit: u64,
    limit_blinding: Scalar,
}

impl Book {
    fn new(limit: u64) -> Self {
        Book {
            auditor: StateAuditor::new(),
            limit,
            limit_blinding: Scalar::random(&mut OsRng),
        }
    }

    /// Walk a sequence of fills, returning the chain and the opening state.
    fn walk(&self, fills: &[i64]) -> (curve25519_dalek::ristretto::RistrettoPoint, Vec<StateStep>) {
        let mut inventory = 0i64;
        let mut blinding = Scalar::random(&mut OsRng);
        let opening = self.auditor.key.commit(&Scalar::ZERO, &blinding);
        let mut steps = Vec::new();
        for (n, filled) in fills.iter().enumerate() {
            let new_blinding = Scalar::random(&mut OsRng);
            let (step, next) = self
                .auditor
                .prove_update(
                    n as u64,
                    inventory,
                    &blinding,
                    *filled,
                    &Scalar::random(&mut OsRng),
                    self.limit,
                    &self.limit_blinding,
                    &new_blinding,
                    &mut OsRng,
                )
                .expect("honest fill inside the limit");
            steps.push(step);
            inventory = next;
            blinding = new_blinding;
        }
        (opening, steps)
    }
}

#[test]
fn an_honest_book_verifies() {
    let book = Book::new(500);
    let (opening, steps) = book.walk(&[100, -40, 250, -300, 60]);
    assert_eq!(
        book.auditor
            .verify_chain(&opening, &steps, &limit_of(&book)),
        Ok(())
    );
}

#[test]
fn the_sign_convention_is_the_one_the_proof_states() {
    // A maker that sells carries a negative position afterwards.
    let book = Book::new(500);
    let (_, steps) = book.walk(&[100]);
    assert_eq!(steps.len(), 1);
    // inventory = 0 - 100 = -100, which is inside the limit, so it proved.
}

#[test]
fn a_fill_that_breaks_the_limit_cannot_be_proved() {
    let book = Book::new(50);
    let mut inventory = 0i64;
    let blinding = Scalar::random(&mut OsRng);
    let err = book
        .auditor
        .prove_update(
            0,
            inventory,
            &blinding,
            400,
            &Scalar::random(&mut OsRng),
            book.limit,
            &book.limit_blinding,
            &Scalar::random(&mut OsRng),
            &mut OsRng,
        )
        .unwrap_err();
    assert!(err.contains("breaks the committed limit"));
    inventory += 0;
    let _ = inventory;
}

#[test]
fn a_replayed_state_breaks_the_chain() {
    let book = Book::new(500);
    let (opening, mut steps) = book.walk(&[100, -40, 250]);
    steps.swap(1, 2);
    assert!(matches!(
        book.auditor
            .verify_chain(&opening, &steps, &limit_of(&book)),
        Err(ChainError::Forked { .. })
    ));
}

#[test]
fn a_step_from_a_different_book_is_a_fork_not_bad_arithmetic() {
    let book = Book::new(500);
    let (opening, mut steps) = book.walk(&[100, -40]);
    let (_, other) = book.walk(&[70, 30]);
    steps[1] = other.into_iter().nth(1).unwrap();
    assert!(matches!(
        book.auditor
            .verify_chain(&opening, &steps, &limit_of(&book)),
        Err(ChainError::Forked { index: 1, .. })
    ));
}

#[test]
fn a_limit_outside_the_ceiling_is_refused_at_commitment() {
    let book = Book::new(500);
    assert!(book
        .auditor
        .commit_limit(u64::MAX, &book.limit_blinding)
        .is_err());
}

#[test]
fn a_chain_is_checked_against_its_own_limit() {
    let book = Book::new(500);
    let (opening, steps) = book.walk(&[400]);
    // The same steps under a tighter committed limit must not verify: the
    // containment proofs are about a difference the tighter limit does not
    // reproduce.
    let tight = Book {
        auditor: StateAuditor::new(),
        limit: 100,
        limit_blinding: book.limit_blinding,
    };
    let tight_limit = tight
        .auditor
        .commit_limit(tight.limit, &tight.limit_blinding)
        .unwrap();
    assert!(book
        .auditor
        .verify_chain(&opening, &steps, &tight_limit)
        .is_err());
}

fn limit_of(book: &Book) -> qomm_proofs::state_audit::InventoryLimit {
    book.auditor
        .commit_limit(book.limit, &book.limit_blinding)
        .unwrap()
}

// --- and the state audited has to be the state the circuit was fed ---------

/// The commitments the input dealer published for each step's opening state.
fn dealt_from(
    opening: &curve25519_dalek::ristretto::RistrettoPoint,
    steps: &[StateStep],
) -> Vec<curve25519_dalek::ristretto::CompressedRistretto> {
    let mut out = Vec::with_capacity(steps.len());
    let mut previous = *opening;
    for step in steps {
        out.push(previous.compress());
        previous = step.inventory;
    }
    out
}

#[test]
fn a_chain_bound_to_what_was_dealt_verifies() {
    let book = Book::new(500);
    let (opening, steps) = book.walk(&[100, -40, 250]);
    let dealt = dealt_from(&opening, &steps);
    assert_eq!(
        book.auditor
            .verify_chain_bound(&opening, &steps, &limit_of(&book), &dealt),
        Ok(())
    );
}

#[test]
fn an_impeccable_chain_over_an_inventory_the_circuit_never_saw_is_refused() {
    // The gap this closes, stated as the test that would have caught it: every
    // proof in the chain verifies, and the chain is about a different book.
    let book = Book::new(500);
    let (opening, steps) = book.walk(&[100, -40, 250]);
    assert_eq!(
        book.auditor
            .verify_chain(&opening, &steps, &limit_of(&book)),
        Ok(())
    );

    let mut dealt = dealt_from(&opening, &steps);
    let elsewhere = book
        .auditor
        .key
        .commit(&Scalar::from(7u64), &Scalar::random(&mut OsRng));
    dealt[1] = elsewhere.compress();
    assert!(matches!(
        book.auditor
            .verify_chain_bound(&opening, &steps, &limit_of(&book), &dealt),
        Err(ChainError::NotTheDealtState { index: 1, .. })
    ));
}

#[test]
fn one_step_can_be_bound_on_its_own() {
    let book = Book::new(500);
    let (opening, steps) = book.walk(&[100, -40]);
    let limit = limit_of(&book);
    assert!(book
        .auditor
        .verify_update_bound(&steps[0], &opening, &limit, &opening.compress()));
    let elsewhere = book
        .auditor
        .key
        .commit(&Scalar::from(3u64), &Scalar::random(&mut OsRng));
    assert!(!book
        .auditor
        .verify_update_bound(&steps[0], &opening, &limit, &elsewhere.compress()));
}

#[test]
fn a_dealt_list_of_the_wrong_length_is_refused_rather_than_zipped() {
    let book = Book::new(500);
    let (opening, steps) = book.walk(&[100, -40, 250]);
    let dealt = dealt_from(&opening, &steps);
    assert!(book
        .auditor
        .verify_chain_bound(&opening, &steps, &limit_of(&book), &dealt[..2])
        .is_err());
}
